repo.go

```diff
@@ -877,3 +1048,315 @@ func (r *Repo) timestampFileMeta(name string) (data.TimestampFileMeta, error) {
+		changed = true
+	}
+
+	if target.TargetAddKey(pk) {
+		changed = true
+	}
+
+	if !changed {
+		return nil
+	}
+
+	target.Expires = expires.Round(time.Second)
+	if _, ok := r.versionUpdated["targets.json"]; !ok {
+		target.Version++
+		r.versionUpdated["targets.json"] = struct{}{}
+	}
+
+	return r.setMeta("targets.json", target)
+}
+
+//DelegateRevokeKey revoke a key for a certain delegated role
+func (r *Repo) DelegateRevokeKey(role, id string) error {
+	return r.DelegateRevokeKeyWithExpires(role, id, data.DefaultExpires("targets"))
+}
+
+//DelegateRevokeKeyWithExpires revoke a key with new expire set
+func (r *Repo) DelegateRevokeKeyWithExpires(keyRole, id string, expires time.Time) error {
+
+	if !validExpires(expires) {
+		return ErrInvalidExpires{expires}
+	}
+
+	targ, err := r.targets()
+	if err != nil {
+		return err
+	}
+
+	key, ok := targ.Keys[id]
+	if !ok {
+		return ErrKeyNotFound{keyRole, id}
+	}
+
+	role, ok := targ.Roles[keyRole]
+	if !ok {
+		return ErrKeyNotFound{keyRole, id}
+	}
+
+	keyIDs := make([]string, 0, len(role.KeyIDs))
+
+	// There may be multiple keyids that correspond to this key, so
+	// filter all of them out.
+	for _, keyID := range role.KeyIDs {
+		if key.ContainsID(keyID) {
+			continue
+		}
+		keyIDs = append(keyIDs, keyID)
+	}
+	if len(keyIDs) == len(role.KeyIDs) {
+		return ErrKeyNotFound{keyRole, id}
+	}
+	role.KeyIDs = keyIDs
+
+	for _, keyID := range key.IDs() {
+		delete(targ.Keys, keyID)
+	}
+	targ.Roles[keyRole] = role
+	targ.Expires = expires.Round(time.Second)
+	if _, ok := r.versionUpdated["targets.json"]; !ok {
+		targ.Version++
+		r.versionUpdated["targets.json"] = struct{}{}
+	}
+
+	return r.setMeta("targets.json", targ)
+}
+
+//DelegateAddTarget add a target to non-top target role
+func (r *Repo) DelegateAddTarget(nameJSON, path string, custom json.RawMessage) error {
+	if !strings.Contains(nameJSON, ".json") {
+		nameJSON = nameJSON + ".json"
+	}
+	return r.DelegateAddTargets(nameJSON, []string{path}, custom)
+}
+
+//DelegateAddTargets add targets to non-top target role
+func (r *Repo) DelegateAddTargets(nameJSON string, paths []string, custom json.RawMessage) error {
+	if !strings.Contains(nameJSON, ".json") {
+		nameJSON = nameJSON + ".json"
+	}
+	return r.DelegateAddTargetsWithExpires(nameJSON, paths, custom, data.DefaultExpires("targets"))
+}
+
+//DelegateAddTargetWithExpires adds single target to non-top target  role
+func (r *Repo) DelegateAddTargetWithExpires(nameJSON string, path string, custom json.RawMessage, expires time.Time) error {
+	if !strings.Contains(nameJSON, ".json") {
+		nameJSON = nameJSON + ".json"
+	}
+	return r.DelegateAddTargetsWithExpires(nameJSON, []string{path}, custom, expires)
+}
+
+//DelegateAddTargetsWithExpires add targets to non-top target role with expire date
+func (r *Repo) DelegateAddTargetsWithExpires(nameJSON string, paths []string, custom json.RawMessage, expires time.Time) error {
+	if !strings.Contains(nameJSON, ".json") {
```

Comment:
I'd also make sure the '.json' is at the end of the string, here and elsewhere in this file